src/ctap/storage/key.rs

```diff
@@ -0,0 +1,135 @@
+macro_rules! make_key {
+    ($(#[$doc: meta])* $name: ident = $key: literal..$end: literal) => {
+        $(#[$doc])* pub const $name: core::ops::Range<usize> = $key..$end;
+    };
+    ($(#[$doc: meta])* $name: ident = $key: literal) => {
+        $(#[$doc])* pub const $name: usize = $key;
+    };
+}
+
+/// Returns the range of values of a key given its value description.
+#[cfg(test)]
+macro_rules! make_range {
+    ($key: literal..$end: literal) => {
+        $key..$end
+    };
+    ($key: literal) => {
+        $key..$key + 1
+    };
+}
+
+/// Helper to define keys as a partial partition of a range.
+macro_rules! make_partition {
+        ($range: expr,
+         $(
+             $(#[$doc: meta])*
+             $name: ident = $key: literal $(.. $end: literal)?;
+         )*) => {
+            $(
+                make_key!($(#[$doc])* $name = $key $(.. $end)?);
+            )*
+            #[cfg(test)]
+            const KEY_RANGE: core::ops::Range<usize> = $range;
+            #[cfg(test)]
+            const ALL_KEYS: &[core::ops::Range<usize>] = &[$(make_range!($key $(.. $end)?)),*];
+        };
+    }
+
+make_partition! {
+    // We reserve 0 and 2048+ for possible migration purposes. We add persistent entries starting
+    // from 1 and going up. We add non-persistent entries starting from 2047 and going down. This
+    // way, we don't commit to a fixed number of persistent keys.
+    1..2048,
+
+    // WARNING: Keys should not be deleted but prefixed with `_` to avoid accidentally reusing them.
+
+    /// The attestation private key.
+    ATTESTATION_PRIVATE_KEY = 1;
+
+    /// The attestation certificate.
+    ATTESTATION_CERTIFICATE = 2;
+
+    /// The aaguid.
+    AAGUID = 3;
+
+    // This is the persistent key limit:
+    // - When adding a (persistent) key above this message, make sure its value is smaller than
+    //   NUM_PERSISTENT_KEYS.
+    // - When adding a (non-persistent) key below this message, make sure its value is bigger or
+    //   equal than NUM_PERSISTENT_KEYS.
+
+    /// The credentials.
+    ///
+    /// Depending on `MAX_SUPPORTED_RESIDENTIAL_KEYS`, only a prefix of those keys is used. Each
+    /// board may configure `MAX_SUPPORTED_RESIDENTIAL_KEYS` depending on the storage size.
+    CREDENTIALS = 1700..2000;
+
+    /// TODO: Help from reviewer needed for documentation.
+    _MIN_PIN_LENGTH_RP_IDS = 2042;
+
+    /// The minimum PIN length.
+    #[cfg(feature = "with_ctap2_1")]
+    MIN_PIN_LENGTH = 2043;
+
+    /// The number of PIN retries.
+    ///
+    /// If the entry is absent, the number of PIN retries is `MAX_PIN_RETRIES`.
+    PIN_RETRIES = 2044;
+
+    /// The PIN hash.
+    ///
+    /// If the entry is absent, there is no PIN set.
+    PIN_HASH = 2045;
+
+    /// The encryption and hmac keys.
+    ///
+    /// This entry is always present. It is generated at startup if absent. This is not a persistent
+    /// key because its value should change after a CTAP reset.
```

Comment:
It depends of which reset we talk about: board reset or CTAP reset. Since we are in the scope of persistent storage, all keys are persistent regarding board reset. The remaining persistence concerns are only about CTAP reset. In particular I usually make this clear in comments by adding "CTAP reset" or "CTAP reset command" when talking about persistence at top-level points like line 15 and even in this line.

Reply:
I see. What is special is that it's always present (other keys can be absent). And I didn't want the reader to think that it's because they persist CTAP reset. But indeed it's actually less confusing to not write it :D Removed.